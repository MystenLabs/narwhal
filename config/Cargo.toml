[package]
name = "config"
version = "0.1.0"
license = "Apache-2.0"
authors = ["Mysten Labs <build@mystenlabs.com>"]
edition = "2021"

[dependencies]
serde = { version = "1.0.139", features = ["derive"] }
serde_json = "1.0.82"
thiserror = "1.0.31"
tracing = "0.1.35"
multiaddr = "0.14.0"

crypto = { path = "../crypto" }
arc-swap = { version = "1.5.0", features = ["serde"] }
workspace-hack = { version = "0.1", path = "../workspace-hack" }

[dev-dependencies]
insta = "1.16.0"
rand = "0.7.3"
tempfile = "3.3.0"
tracing-test = "0.2.3"

test_utils = { path = "../test_utils"}
