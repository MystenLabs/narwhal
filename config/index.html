<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `config` crate."><meta name="keywords" content="rust, rustlang, rust-lang, config"><title>config - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../ayu.css" disabled><link rel="stylesheet" type="text/css" href="../dark.css" disabled><link rel="stylesheet" type="text/css" href="../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../storage.js"></script><script defer src="../crates.js"></script><script defer src="../main.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../config/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div>
        </a><h2 class="location"></h2>
    </nav>
    <nav class="sidebar"><a class="sidebar-logo" href="../config/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div>
        </a><h2 class="location"><a href="#">Crate config</a></h2><div class="sidebar-elems"><div class="block"><ul><li class="version">Version 0.1.0</li><li><a id="all-types" href="all.html">All Items</a></li></ul></div><section><div class="block"><ul><li><a href="#modules">Modules</a></li><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#traits">Traits</a></li><li><a href="#types">Type Definitions</a></li></ul></div></section></div></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../config/index.html"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></a><nav class="sub"><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><button type="button">?</button></div><div id="settings-menu" tabindex="-1">
                                <a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../wheel.svg"></a></div>
                        </div></form></nav></div><section id="main-content" class="content"><div class="main-heading">
    <h1 class="fqn"><span class="in-band">Crate <a class="mod" href="#">config</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span></h1><span class="out-of-band"><a class="srclink" href="../src/config/lib.rs.html#4-659">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><h2 id="modules" class="small-section-header"><a href="#modules">Modules</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="mod" href="utils/index.html" title="config::utils mod">utils</a></div><div class="item-right docblock-short"></div></div></div><h2 id="structs" class="small-section-header"><a href="#structs">Structs</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.Authority.html" title="config::Authority struct">Authority</a></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.BlockSynchronizerParameters.html" title="config::BlockSynchronizerParameters struct">BlockSynchronizerParameters</a></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.Committee.html" title="config::Committee struct">Committee</a></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.ConsensusAPIGrpcParameters.html" title="config::ConsensusAPIGrpcParameters struct">ConsensusAPIGrpcParameters</a></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.Parameters.html" title="config::Parameters struct">Parameters</a></div><div class="item-right docblock-short"><p>Holds all the node properties. An example is provided to
showcase the usage and deserialization from a json file.
To define a Duration on the property file can use either
miliseconds or seconds (e.x 5s, 10ms , 2000ms).</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.PrimaryAddresses.html" title="config::PrimaryAddresses struct">PrimaryAddresses</a></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.PrometheusMetricsParameters.html" title="config::PrometheusMetricsParameters struct">PrometheusMetricsParameters</a></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.WorkerAddresses.html" title="config::WorkerAddresses struct">WorkerAddresses</a></div><div class="item-right docblock-short"></div></div></div><h2 id="enums" class="small-section-header"><a href="#enums">Enums</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.ComitteeUpdateError.html" title="config::ComitteeUpdateError enum">ComitteeUpdateError</a></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.ConfigError.html" title="config::ConfigError enum">ConfigError</a></div><div class="item-right docblock-short"></div></div></div><h2 id="traits" class="small-section-header"><a href="#traits">Traits</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.Export.html" title="config::Export trait">Export</a></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.Import.html" title="config::Import trait">Import</a></div><div class="item-right docblock-short"></div></div></div><h2 id="types" class="small-section-header"><a href="#types">Type Definitions</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="type" href="type.Epoch.html" title="config::Epoch type">Epoch</a></div><div class="item-right docblock-short"><p>The epoch number.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="type" href="type.SharedCommittee.html" title="config::SharedCommittee type">SharedCommittee</a></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left module-item"><a class="type" href="type.Stake.html" title="config::Stake type">Stake</a></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left module-item"><a class="type" href="type.WorkerId.html" title="config::WorkerId type">WorkerId</a></div><div class="item-right docblock-short"></div></div></div></section></div></main><div id="rustdoc-vars" data-root-path="../" data-current-crate="config" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.64.0-nightly (6dbae3ad1 2022-07-25)" ></div>
</body></html>