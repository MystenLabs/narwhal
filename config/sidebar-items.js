window.SIDEBAR_ITEMS = {"enum":[["ComitteeUpdateError",""],["ConfigError",""]],"mod":[["utils",""]],"struct":[["Authority",""],["BlockSynchronizerParameters",""],["Committee",""],["ConsensusAPIGrpcParameters",""],["Parameters","Holds all the node properties. An example is provided to showcase the usage and deserialization from a json file. To define a Duration on the property file can use either miliseconds or seconds (e.x 5s, 10ms , 2000ms)."],["PrimaryAddresses",""],["PrometheusMetricsParameters",""],["WorkerAddresses",""]],"trait":[["Export",""],["Import",""]],"type":[["Epoch","The epoch number."],["SharedCommittee",""],["Stake",""],["WorkerId",""]]};