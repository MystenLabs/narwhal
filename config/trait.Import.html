<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `Import` trait in crate `config`."><meta name="keywords" content="rust, rustlang, rust-lang, Import"><title>Import in config - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../ayu.css" disabled><link rel="stylesheet" type="text/css" href="../dark.css" disabled><link rel="stylesheet" type="text/css" href="../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../storage.js"></script><script defer src="sidebar-items.js"></script><script defer src="../main.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"></head><body class="rustdoc trait"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../config/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div>
        </a><h2 class="location"></h2>
    </nav>
    <nav class="sidebar"><a class="sidebar-logo" href="../config/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div>
        </a><h2 class="location"><a href="#">Import</a></h2><div class="sidebar-elems"><section><div class="block"><h3 class="sidebar-title"><a href="#provided-methods">Provided Methods</a></h3><ul><li><a href="#method.import">import</a></li></ul></div><div class="block"><h3 class="sidebar-title"><a href="#implementors">Implementors</a></h3></div></section><h2 class="location"><a href="index.html">In config</a></h2></div></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../config/index.html"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></a><nav class="sub"><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><button type="button">?</button></div><div id="settings-menu" tabindex="-1">
                                <a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../wheel.svg"></a></div>
                        </div></form></nav></div><section id="main-content" class="content"><div class="main-heading">
    <h1 class="fqn"><span class="in-band">Trait <a href="index.html">config</a>::<wbr><a class="trait" href="#">Import</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span></h1><span class="out-of-band"><a class="srclink" href="../src/config/lib.rs.html#61-72">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><div class="docblock item-decl"><pre class="rust trait"><code>pub trait Import: <a class="trait" href="https://docs.rs/serde/1.0.140/serde/de/trait.DeserializeOwned.html" title="trait serde::de::DeserializeOwned">DeserializeOwned</a> {
    fn <a href="#method.import" class="fnname">import</a>(path: &amp;<a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.str.html">str</a>) -&gt; <a class="enum" href="https://doc.rust-lang.org/nightly/core/result/enum.Result.html" title="enum core::result::Result">Result</a>&lt;Self, <a class="enum" href="enum.ConfigError.html" title="enum config::ConfigError">ConfigError</a>&gt; { ... }
}</code></pre></div><h2 id="provided-methods" class="small-section-header">Provided Methods<a href="#provided-methods" class="anchor"></a></h2><div class="methods"><div id="method.import" class="method has-srclink"><div class="rightside"><a class="srclink" href="../src/config/lib.rs.html#62-71">source</a></div><h4 class="code-header">fn <a href="#method.import" class="fnname">import</a>(path: &amp;<a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.str.html">str</a>) -&gt; <a class="enum" href="https://doc.rust-lang.org/nightly/core/result/enum.Result.html" title="enum core::result::Result">Result</a>&lt;Self, <a class="enum" href="enum.ConfigError.html" title="enum config::ConfigError">ConfigError</a>&gt;</h4></div></div><h2 id="implementors" class="small-section-header">Implementors<a href="#implementors" class="anchor"></a></h2><div class="item-list" id="implementors-list"><section id="impl-Import-for-D" class="impl has-srclink"><span class="rightside"><a class="srclink" href="../src/config/lib.rs.html#74">source</a></span><a href="#impl-Import-for-D" class="anchor"></a><h3 class="code-header in-band">impl&lt;D:&nbsp;<a class="trait" href="https://docs.rs/serde/1.0.140/serde/de/trait.DeserializeOwned.html" title="trait serde::de::DeserializeOwned">DeserializeOwned</a>&gt; <a class="trait" href="trait.Import.html" title="trait config::Import">Import</a> for D</h3></section></div><script type="text/javascript" src="../implementors/config/trait.Import.js" data-ignore-extern-crates="" async></script></section></div></main><div id="rustdoc-vars" data-root-path="../" data-current-crate="config" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.64.0-nightly (6dbae3ad1 2022-07-25)" ></div>
</body></html>