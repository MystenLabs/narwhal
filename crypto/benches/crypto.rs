// Copyright (c) 2022, Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0
#[macro_use]
extern crate criterion;
extern crate ed25519_dalek;
extern crate rand;

mod ed25519_benches {
    use super::*;
    use blake2::digest::Update;
    use criterion::*;
    use crypto::{
        bls12377::{BLS12377KeyPair, BLS12377Signature},
        bls12381::{BLS12381KeyPair, BLS12381Signature},
        ed25519::*,
        secp256k1::{Secp256k1KeyPair, Secp256k1Signature},
        traits::{KeyPair, VerifyingKey},
        Verifier,
    };
    use rand::{prelude::ThreadRng, thread_rng};
    use signature::Signer;

    fn sign(c: &mut Criterion) {
        let mut csprng: ThreadRng = thread_rng();
        let ed_keypair = Ed25519KeyPair::generate(&mut csprng);
        let bls_keypair = BLS12377KeyPair::generate(&mut csprng);
        let blst_keypair = BLS12381KeyPair::generate(&mut csprng);
        let secp256k1_keypair = Secp256k1KeyPair::generate(&mut csprng);
        let msg: &[u8] = b"";

        c.bench_function("Ed25519 signing", move |b| b.iter(|| ed_keypair.sign(msg)));
        c.bench_function("BLS12377 signing", move |b| {
            b.iter(|| bls_keypair.sign(msg))
        });
        c.bench_function("BLS12381 signing", move |b| {
            b.iter(|| blst_keypair.sign(msg))
        });
        c.bench_function("Secp256k1 signing", move |b| {
            b.iter(|| secp256k1_keypair.sign(msg))
        });
    }

    fn verify(c: &mut Criterion) {
        let mut csprng: ThreadRng = thread_rng();
        let ed_keypair = Ed25519KeyPair::generate(&mut csprng);
        let bls_keypair = BLS12377KeyPair::generate(&mut csprng);
        let blst_keypair = BLS12381KeyPair::generate(&mut csprng);
        let secp256k1_keypair = Secp256k1KeyPair::generate(&mut csprng);

        let ed_public = ed_keypair.public();
        let bls_public = bls_keypair.public();
        let blst_public = blst_keypair.public();
        let secp256k1_public = secp256k1_keypair.public();

        let msg: &[u8] = b"";
        let ed_sig: Ed25519Signature = ed_keypair.sign(msg);
        let bls_sig: BLS12377Signature = bls_keypair.sign(msg);
        let blst_sig: BLS12381Signature = blst_keypair.sign(msg);
        let secp256k1_sig: Secp256k1Signature = secp256k1_keypair.sign(msg);

        c.bench_function("Ed25519 signature verification", move |b| {
            b.iter(|| ed_public.verify(msg, &ed_sig))
        });
        c.bench_function("BLS12377 signature verification", move |b| {
            b.iter(|| bls_public.verify(msg, &bls_sig))
        });
        c.bench_function("BLS12381 signature verification", move |b| {
            b.iter(|| blst_public.verify(msg, &blst_sig))
        });
        c.bench_function("Secp256k1 signature verification", move |b| {
            b.iter(|| secp256k1_public.verify(msg, &secp256k1_sig))
        });
    }

    fn verify_batch_signatures<M: measurement::Measurement>(c: &mut BenchmarkGroup<M>) {
        static BATCH_SIZES: [usize; 10] = [16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192];

        let mut csprng: ThreadRng = thread_rng();

        for size in BATCH_SIZES.iter() {
            let ed_keypairs: Vec<_> = (0..*size)
                .map(|_| Ed25519KeyPair::generate(&mut csprng))
                .collect();
            let bls_keypairs: Vec<_> = (0..*size)
                .map(|_| BLS12377KeyPair::generate(&mut csprng))
                .collect();
            let blst_keypairs: Vec<_> = (0..*size)
                .map(|_| BLS12381KeyPair::generate(&mut csprng))
                .collect();

            let msg: Vec<u8> = {
                crypto::blake2b_256(|hasher| {
                    hasher.update(b"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")
                })
                .to_vec()
            };

            let ed_signatures: Vec<_> = ed_keypairs.iter().map(|key| key.sign(&msg)).collect();
            let ed_public_keys: Vec<_> =
                ed_keypairs.iter().map(|key| key.public().clone()).collect();
            let bls_signatures: Vec<_> = bls_keypairs.iter().map(|key| key.sign(&msg)).collect();
            let bls_public_keys: Vec<_> = bls_keypairs
                .iter()
                .map(|key| key.public().clone())
                .collect();
            let blst_signatures: Vec<_> = blst_keypairs.iter().map(|key| key.sign(&msg)).collect();
            let blst_public_keys: Vec<_> = blst_keypairs
                .iter()
                .map(|key| key.public().clone())
                .collect();

            c.bench_with_input(
                BenchmarkId::new("Ed25519 batch verification", *size),
                &(msg.clone(), ed_public_keys, ed_signatures),
                |b, i| {
                    b.iter(|| VerifyingKey::verify_batch(&i.0, &i.1[..], &i.2[..]));
                },
            );
            c.bench_with_input(
                BenchmarkId::new("BLS12377 batch verification", *size),
                &(msg.clone(), bls_public_keys, bls_signatures),
                |b, i| {
                    b.iter(|| VerifyingKey::verify_batch(&i.0, &i.1[..], &i.2[..]));
                },
            );
            c.bench_with_input(
                BenchmarkId::new("BLS12381 batch verification", *size),
                &(msg, blst_public_keys, blst_signatures),
                |b, i| {
                    b.iter(|| VerifyingKey::verify_batch(&i.0, &i.1[..], &i.2[..]));
                },
            );
        }
    }

    fn key_generation(c: &mut Criterion) {
        let mut csprng: ThreadRng = thread_rng();

        c.bench_function("Ed25519 keypair generation", move |b| {
            b.iter(|| Ed25519KeyPair::generate(&mut csprng))
        });
        c.bench_function("BLS12377 keypair generation", move |b| {
            b.iter(|| BLS12377KeyPair::generate(&mut csprng))
        });
        c.bench_function("BLS12381 keypair generation", move |b| {
            b.iter(|| BLS12381KeyPair::generate(&mut csprng))
        });
        c.bench_function("Secp256k1 keypair generation", move |b| {
            b.iter(|| Secp256k1KeyPair::generate(&mut csprng))
        });
    }

    criterion_group! {
        name = ed25519_benches;
        config = Criterion::default();
        targets =
           sign,
           verify,
           verification_comparison,
           key_generation
    }

    fn verification_comparison(c: &mut Criterion) {
        let mut group: BenchmarkGroup<_> = c.benchmark_group("verification_comparison");
        group.sampling_mode(SamplingMode::Flat);

        verify_batch_signatures(&mut group);
        group.finish();
    }
}

criterion_main!(ed25519_benches::ed25519_benches,);
