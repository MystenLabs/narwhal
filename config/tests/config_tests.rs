// Copyright (c) 2022, Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use std::collections::BTreeMap;

use config::{
    Committee, ConsensusAPIGrpcParameters, Epoch, Parameters, PrimaryAddresses,
    PrometheusMetricsParameters, Stake,
};
use crypto::{ed25519::Ed25519PublicKey, traits::KeyPair};
use insta::assert_json_snapshot;
use rand::seq::SliceRandom;
use test_utils::make_authority_with_port_getter;

#[test]
fn update_primary_network_info_test() {
    let committee = test_utils::committee(None);
    let res = committee
        .clone()
        .update_primary_network_info(BTreeMap::new())
        .unwrap_err();
    for err in res {
        assert!(matches!(
            err,
            config::ComitteeUpdateError::MissingFromUpdate(_)
        ))
    }

    let committee2 = test_utils::committee(42);
    let invalid_new_info = committee2
        .authorities
        .iter()
        .map(|(pk, a)| (pk.clone(), (a.stake, a.primary.clone())))
        .collect::<BTreeMap<_, (Stake, PrimaryAddresses)>>();
    let res2 = committee
        .clone()
        .update_primary_network_info(invalid_new_info)
        .unwrap_err();
    for err in res2 {
        // we'll get the two collections reporting missing from each other
        assert!(matches!(
            err,
            config::ComitteeUpdateError::NotInCommittee(_)
                | config::ComitteeUpdateError::MissingFromUpdate(_)
        ))
    }

    let committee3 = test_utils::committee(None);
    let invalid_new_info = committee3
        .authorities
        .iter()
        // change the stake
        .map(|(pk, a)| (pk.clone(), (a.stake + 1, a.primary.clone())))
        .collect::<BTreeMap<_, (Stake, PrimaryAddresses)>>();
    let res2 = committee
        .clone()
        .update_primary_network_info(invalid_new_info)
        .unwrap_err();
    for err in res2 {
        assert!(matches!(
            err,
            config::ComitteeUpdateError::DifferentStake(_)
        ))
    }

    let committee4 = test_utils::committee(None);
    let mut pk_n_stake = Vec::new();
    let mut addresses = Vec::new();

    committee4.authorities.iter().for_each(|(pk, a)| {
        pk_n_stake.push((pk.clone(), a.stake));
        addresses.push(a.primary.clone())
    });

    let mut rng = rand::thread_rng();
    addresses.shuffle(&mut rng);

    let new_info = pk_n_stake
        .into_iter()
        .zip(addresses)
        .map(|((pk, stk), addr)| (pk, (stk, addr)))
        .collect::<BTreeMap<Ed25519PublicKey, (Stake, PrimaryAddresses)>>();

    let mut comm = committee;
    let res = comm.update_primary_network_info(new_info.clone());
    assert!(res.is_ok());
    for (pk, a) in comm.authorities.iter() {
        assert_eq!(a.primary, new_info.get(pk).unwrap().1);
    }
}

#[test]
fn parameters_snapshot_matches() {
    // This configuration is load-bearing in the NW benchmarks,
    // and in Sui (prod config + shared object bench base). If this test breaks,
    // config needs to change in all of these.

    // We avoid defaults that randomly bind to a random port
    let consensus_api_grpc_parameters = ConsensusAPIGrpcParameters {
        socket_addr: "/ip4/127.0.0.1/tcp/8081/http".parse().unwrap(),
        ..ConsensusAPIGrpcParameters::default()
    };
    let prometheus_metrics_parameters = PrometheusMetricsParameters {
        socket_addr: "127.0.0.1:8081".parse().unwrap(),
    };

    let parameters = Parameters {
        consensus_api_grpc: consensus_api_grpc_parameters,
        prometheus_metrics: prometheus_metrics_parameters,
        ..Parameters::default()
    };
    assert_json_snapshot!("parameters", parameters)
}

#[test]
fn commmittee_snapshot_matches() {
    // The shape of this configuration is load-bearing in the NW benchmarks,
    // and in Sui (prod)
    let keys = test_utils::keys(None);

    let committee = Committee {
        epoch: Epoch::default(),
        authorities: keys
            .iter()
            .map(|kp| {
                let mut port = 0;
                let increment_port_getter = || {
                    port += 1;
                    port
                };
                (
                    kp.public().clone(),
                    make_authority_with_port_getter(increment_port_getter),
                )
            })
            .collect(),
    };
    // we need authorities to be serialized in order
    let mut settings = insta::Settings::clone_current();
    settings.set_sort_maps(true);
    settings.bind(|| assert_json_snapshot!("committee", committee));
}
