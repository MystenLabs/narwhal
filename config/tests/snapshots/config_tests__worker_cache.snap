---
source: config/tests/config_tests.rs
expression: worker_cache
---
{
  "workers": {
    "O6EMZUvkFG2rPW2rgLPOJM4323x0XCslMgXWL5Z7ALM=": {
      "0": {
        "name": "dt4/oHrYlTWhdms3mHV8irOu9b6H3uFunhBt2LUJcp4=",
        "transactions": "/ip4/127.0.0.1/tcp/0/http",
        "worker_to_worker": "/ip4/127.0.0.1/tcp/0/http",
        "primary_to_worker": "/ip4/127.0.0.1/tcp/0/http"
      },
      "1": {
        "name": "lrisYqEadH5cgLhsT1EW5q6uVKxwCXLTJQO3klahUy8=",
        "transactions": "/ip4/127.0.0.1/tcp/0/http",
        "worker_to_worker": "/ip4/127.0.0.1/tcp/0/http",
        "primary_to_worker": "/ip4/127.0.0.1/tcp/0/http"
      },
      "2": {
        "name": "VayIYzPsiroVrY4eilEUVKCM/LfNcMWN2ELKoHCNGr8=",
        "transactions": "/ip4/127.0.0.1/tcp/0/http",
        "worker_to_worker": "/ip4/127.0.0.1/tcp/0/http",
        "primary_to_worker": "/ip4/127.0.0.1/tcp/0/http"
      },
      "3": {
        "name": "Kt9mRluFkBunwfq2VREQbBXuSYGOsFo95bA/PIrvVhc=",
        "transactions": "/ip4/127.0.0.1/tcp/0/http",
        "worker_to_worker": "/ip4/127.0.0.1/tcp/0/http",
        "primary_to_worker": "/ip4/127.0.0.1/tcp/0/http"
      }
    },
    "gznZnDM7s6J3/reAuNRsrDDNzZBQ2+I1+4zGe5HBsGo=": {
      "0": {
        "name": "hkGX9Mxcq2yO3Ut5xIXvuaVVcYt1tgPi6LmPVewN4mc=",
        "transactions": "/ip4/127.0.0.1/tcp/0/http",
        "worker_to_worker": "/ip4/127.0.0.1/tcp/0/http",
        "primary_to_worker": "/ip4/127.0.0.1/tcp/0/http"
      },
      "1": {
        "name": "vVDfe4JyrKvldWAMAgdqeUvACFqw15e4PUoj4ByOfCQ=",
        "transactions": "/ip4/127.0.0.1/tcp/0/http",
        "worker_to_worker": "/ip4/127.0.0.1/tcp/0/http",
        "primary_to_worker": "/ip4/127.0.0.1/tcp/0/http"
      },
      "2": {
        "name": "+oKq7VFT0pAQvjlvEti2YUw5fHXEYT/bs21P9wTlAKs=",
        "transactions": "/ip4/127.0.0.1/tcp/0/http",
        "worker_to_worker": "/ip4/127.0.0.1/tcp/0/http",
        "primary_to_worker": "/ip4/127.0.0.1/tcp/0/http"
      },
      "3": {
        "name": "FsiKyAGnhMcChNBf+L2B8EHX6mMup8oqJ8bK/9ZRkww=",
        "transactions": "/ip4/127.0.0.1/tcp/0/http",
        "worker_to_worker": "/ip4/127.0.0.1/tcp/0/http",
        "primary_to_worker": "/ip4/127.0.0.1/tcp/0/http"
      }
    },
    "nUdhahjfIDkwfwZCImAliLUqaq6qe2/oxrTZ66inh7E=": {
      "0": {
        "name": "6j/Odu4Qs6ZF+zBAUR36tMnNYFj0GSkBTRZGb147LCo=",
        "transactions": "/ip4/127.0.0.1/tcp/0/http",
        "worker_to_worker": "/ip4/127.0.0.1/tcp/0/http",
        "primary_to_worker": "/ip4/127.0.0.1/tcp/0/http"
      },
      "1": {
        "name": "r7SQXpVQz0YqggB4JKBV2XWuk0RwxHfZ3bttKyZMNvo=",
        "transactions": "/ip4/127.0.0.1/tcp/0/http",
        "worker_to_worker": "/ip4/127.0.0.1/tcp/0/http",
        "primary_to_worker": "/ip4/127.0.0.1/tcp/0/http"
      },
      "2": {
        "name": "l+fthD+rGkLOnCEpQ1IewioCVodHzAA7iB+9J/cL3+A=",
        "transactions": "/ip4/127.0.0.1/tcp/0/http",
        "worker_to_worker": "/ip4/127.0.0.1/tcp/0/http",
        "primary_to_worker": "/ip4/127.0.0.1/tcp/0/http"
      },
      "3": {
        "name": "XEf+9+MRPcAcVu9NR3v+UVevwzlpoBT4u1U34YzOxkw=",
        "transactions": "/ip4/127.0.0.1/tcp/0/http",
        "worker_to_worker": "/ip4/127.0.0.1/tcp/0/http",
        "primary_to_worker": "/ip4/127.0.0.1/tcp/0/http"
      }
    },
    "ucbuFjDvPnERRKZI2wa7sihPcnTPvuU//O5QPMGkkgA=": {
      "0": {
        "name": "rvP0pLjsod/DQzYb+OQ2vULeklnAS4MU644gVN1ugqs=",
        "transactions": "/ip4/127.0.0.1/tcp/0/http",
        "worker_to_worker": "/ip4/127.0.0.1/tcp/0/http",
        "primary_to_worker": "/ip4/127.0.0.1/tcp/0/http"
      },
      "1": {
        "name": "ildi4hrBzbOHBELHe0w69Yx87bh3nQJw5tTx4vc2fXQ=",
        "transactions": "/ip4/127.0.0.1/tcp/0/http",
        "worker_to_worker": "/ip4/127.0.0.1/tcp/0/http",
        "primary_to_worker": "/ip4/127.0.0.1/tcp/0/http"
      },
      "2": {
        "name": "zGIzLjS7LVzWn2Dvuyo2y5FsfrRYMB6jZjbE27ASvYg=",
        "transactions": "/ip4/127.0.0.1/tcp/0/http",
        "worker_to_worker": "/ip4/127.0.0.1/tcp/0/http",
        "primary_to_worker": "/ip4/127.0.0.1/tcp/0/http"
      },
      "3": {
        "name": "I47bbumpH9yOcb5QwKP7PPKS2QJN26wCqqcJCWsKqdo=",
        "transactions": "/ip4/127.0.0.1/tcp/0/http",
        "worker_to_worker": "/ip4/127.0.0.1/tcp/0/http",
        "primary_to_worker": "/ip4/127.0.0.1/tcp/0/http"
      }
    }
  },
  "epoch": 0
}
